File: java/src/test/java/co/rerout/sdk/SmartLinksTest.java
/*
 * rerout-java — Smart Links surface tests.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import co.rerout.sdk.MockApi.MockResponse;
import co.rerout.sdk.MockApi.RecordedRequest;
import co.rerout.sdk.model.AbVariant;
import co.rerout.sdk.model.AbVariantInput;
import co.rerout.sdk.model.BatchCreateLinksResult;
import co.rerout.sdk.model.BatchLinkInput;
import co.rerout.sdk.model.CreateLinkInput;
import co.rerout.sdk.model.Link;
import co.rerout.sdk.model.RoutingRule;
import co.rerout.sdk.model.UpdateLinkInput;
import java.util.List;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SmartLinksTest {

    private MockApi api;

    @BeforeEach
    void setUp() {
        api = new MockApi();
    }

    @AfterEach
    void tearDown() {
        api.close();
    }

    // ─── Link response parsing ──────────────────────────────────────────────

    @Test
    void linkParsesSmartLinkFields() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        Link link = api.client().links().get("q4");

        assertTrue(link.isPasswordProtected());
        assertEquals(Long.valueOf(1000), link.getMaxClicks());
        assertEquals(42L, link.getClickCount());
        assertTrue(link.isTrackConversions());

        assertEquals(1, link.getRoutingRules().size());
        RoutingRule rule = link.getRoutingRules().get(0);
        assertEquals("country", rule.getConditionType());
        assertEquals("is", rule.getConditionOp());
        assertEquals("ZA", rule.getConditionValue());
        assertEquals("https://example.com/za", rule.getTargetUrl());

        assertEquals(2, link.getAbVariants().size());
        AbVariant first = link.getAbVariants().get(0);
        assertEquals(1, first.getId());
        assertEquals("https://example.com/a", first.getTargetUrl());
        assertEquals(70, first.getWeight());
    }

    @Test
    void linkDefaultsSmartLinkFieldsWhenAbsent() {
        // The classic (pre-Smart-Links) payload has none of the new fields.
        api.enqueue(MockResponse.json(MockApi.SAMPLE_LINK_NO_TAGS_JSON));
        Link link = api.client().links().get("q4");

        assertFalse(link.isPasswordProtected());
        assertNull(link.getMaxClicks());
        assertEquals(0L, link.getClickCount());
        assertFalse(link.isTrackConversions());
        assertTrue(link.getRoutingRules().isEmpty());
        assertTrue(link.getAbVariants().isEmpty());
    }

    // ─── CreateLinkInput serialization ──────────────────────────────────────

    @Test
    void createOmitsSmartLinkFieldsWhenUnset() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        api.client().links().create(CreateLinkInput.builder("https://example.com").build());

        String body = api.takeRequest().body;
        assertFalse(body.contains("password"));
        assertFalse(body.contains("max_clicks"));
        assertFalse(body.contains("track_conversions"));
        assertFalse(body.contains("routing_rules"));
        assertFalse(body.contains("ab_variants"));
    }

    @Test
    void createSerializesSmartLinkFields() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        api.client().links().create(
                CreateLinkInput.builder("https://example.com")
                        .password("s3cret")
                        .maxClicks(500)
                        .trackConversions(true)
                        .routingRules(new RoutingRule("device", "in", "mobile,tablet",
                                "https://example.com/m"))
                        .abVariants(
                                new AbVariantInput("https://example.com/a", 60),
                                new AbVariantInput("https://example.com/b"))
                        .build());

        String body = api.takeRequest().body;
        assertTrue(body.contains("\"password\":\"s3cret\""));
        assertTrue(body.contains("\"max_clicks\":500"));
        assertTrue(body.contains("\"track_conversions\":true"));
        assertTrue(body.contains("\"condition_type\":\"device\""));
        assertTrue(body.contains("\"condition_op\":\"in\""));
        assertTrue(body.contains("\"condition_value\":\"mobile,tablet\""));
        assertTrue(body.contains("\"target_url\":\"https://example.com/m\""));
        assertTrue(body.contains("\"target_url\":\"https://example.com/a\""));
        assertTrue(body.contains("\"weight\":60"));
        // The second variant omits weight (server default).
        assertTrue(body.contains("\"target_url\":\"https://example.com/b\""));
    }

    // ─── UpdateLinkInput serialization ──────────────────────────────────────

    @Test
    void updateSerializesSmartLinkFields() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        api.client().links().update("q4",
                UpdateLinkInput.builder()
                        .password("newpass")
                        .maxClicks(250)
                        .trackConversions(true)
                        .routingRules(new RoutingRule("country", "is_not", "US",
                                "https://example.com/intl"))
                        .abVariants(new AbVariantInput("https://example.com/a", 50))
                        .build());

        String body = api.takeRequest().body;
        assertTrue(body.contains("\"password\":\"newpass\""));
        assertTrue(body.contains("\"max_clicks\":250"));
        assertTrue(body.contains("\"track_conversions\":true"));
        assertTrue(body.contains("\"condition_type\":\"country\""));
        assertTrue(body.contains("\"target_url\":\"https://example.com/a\""));
    }

    @Test
    void updateClearsPasswordAndMaxClicks() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        api.client().links().update("q4",
                UpdateLinkInput.builder()
                        .clearPassword()
                        .clearMaxClicks()
                        .build());

        String body = api.takeRequest().body;
        assertTrue(body.contains("\"password\":null"));
        assertTrue(body.contains("\"max_clicks\":null"));
    }

    @Test
    void updateAllowsEmptyRoutingRulesAsFullReplace() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        api.client().links().update("q4",
                UpdateLinkInput.builder().routingRules(List.of()).build());

        String body = api.takeRequest().body;
        assertTrue(body.contains("\"routing_rules\":[]"));
    }

    @Test
    void updateLeavesSmartLinkFieldsAloneByDefault() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_SMART_LINK_JSON));
        api.client().links().update("q4", UpdateLinkInput.builder().active(true).build());

        String body = api.takeRequest().body;
        assertEquals("{\"is_active\":true}", body);
    }

    @Test
    void updateIsEmptyAccountsForSmartLinkFields() {
        assertFalse(UpdateLinkInput.builder().password("x").build().isEmpty());
        assertFalse(UpdateLinkInput.builder().maxClicks(10).build().isEmpty());
        assertFalse(UpdateLinkInput.builder().trackConversions(true).build().isEmpty());
        assertFalse(UpdateLinkInput.builder().routingRules(List.of()).build().isEmpty());
        assertFalse(UpdateLinkInput.builder().abVariants(List.of()).build().isEmpty());
    }

    // ─── createBatch ────────────────────────────────────────────────────────

    @Test
    void createBatchPostsToBatchEndpoint() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_BATCH_RESULT_JSON));
        api.client().links().createBatch(List.of(
                BatchLinkInput.builder("https://example.com/a").build(),
                BatchLinkInput.builder("https://example.com/b").build()));

        RecordedRequest recorded = api.takeRequest();
        assertEquals("POST", recorded.method);
        assertEquals("/v1/links/batch", recorded.path);
    }

    @Test
    void createBatchSerializesLinksArray() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_BATCH_RESULT_JSON));
        api.client().links().createBatch(List.of(
                BatchLinkInput.builder("https://example.com/a")
                        .code("a")
                        .expiresAt(1900000000L)
                        .domainHostname("go.brand.com")
                        .build(),
                BatchLinkInput.builder("https://example.com/b").build()));

        String body = api.takeRequest().body;
        assertTrue(body.contains("\"links\":["));
        assertTrue(body.contains("\"target_url\":\"https://example.com/a\""));
        assertTrue(body.contains("\"code\":\"a\""));
        assertTrue(body.contains("\"expires_at\":1900000000"));
        assertTrue(body.contains("\"domain_hostname\":\"go.brand.com\""));
        assertTrue(body.contains("\"target_url\":\"https://example.com/b\""));
    }

    @Test
    void createBatchParsesResult() {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_BATCH_RESULT_JSON));
        BatchCreateLinksResult result = api.client().links().createBatch(List.of(
                BatchLinkInput.builder("https://example.com/a").build(),
                BatchLinkInput.builder("https://bad").build()));

        assertEquals(1, result.getCreated());
        assertEquals(2, result.getTotal());
        assertEquals(2, result.getResults().size());

        assertEquals(0, result.getResults().get(0).getIndex());
        assertTrue(result.getResults().get(0).isOk());
        assertEquals("aaa", result.getResults().get(0).getCode());
        assertNull(result.getResults().get(0).getError());

        assertEquals(1, result.getResults().get(1).getIndex());
        assertFalse(result.getResults().get(1).isOk());
        assertNull(result.getResults().get(1).getCode());
        assertEquals("bad_target_url", result.getResults().get(1).getError());
    }

    @Test
    void createBatchSurfacesAServerError() {
        api.enqueue(MockResponse.json(
                "{\"code\":\"rate_limited\",\"message\":\"slow down\"}", 429));
        ReroutException ex = assertThrows(ReroutException.class,
                () -> api.client().links().createBatch(List.of(
                        BatchLinkInput.builder("https://example.com/a").build())));
        assertEquals("rate_limited", ex.getCode());
        assertEquals(429, ex.getStatus());
    }

    @Test
    void createBatchAsyncCompletesWithTheParsedResult() throws Exception {
        api.enqueue(MockResponse.json(MockApi.SAMPLE_BATCH_RESULT_JSON));
        BatchCreateLinksResult result = api.client().links().createBatchAsync(List.of(
                BatchLinkInput.builder("https://example.com/a").build())).get();
        assertEquals(1, result.getCreated());
    }
}

File: kotlin/src/test/kotlin/co/rerout/sdk/SmartLinksTest.kt
/*
 * rerout-kotlin — Smart Links field + batch create tests.
 *
 * Copyright (c) 2026 Codecraft Solutions
 * Licensed under the MIT License — https://codecraftsolutions.co.za
 */

package co.rerout.sdk

import kotlinx.coroutines.test.runTest
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class SmartLinksTest {
    private lateinit var server: MockWebServer

    @BeforeEach
    fun setUp() {
        server = MockWebServer()
        server.start()
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    // ─── Link response parsing ────────────────────────────────────────────────

    @Test
    fun `link parses the smart-links fields`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_SMART_LINK_JSON))
        val link = server.client().links.get("vip")
        assertTrue(link.passwordProtected)
        assertEquals(1000L, link.maxClicks)
        assertEquals(42L, link.clickCount)
        assertTrue(link.trackConversions)
        assertEquals(2, link.routingRules.size)
        assertEquals("country", link.routingRules[0].conditionType)
        assertEquals("is", link.routingRules[0].conditionOp)
        assertEquals("ZA", link.routingRules[0].conditionValue)
        assertEquals("https://example.com/za", link.routingRules[0].targetUrl)
        assertEquals("device", link.routingRules[1].conditionType)
        assertEquals("in", link.routingRules[1].conditionOp)
        assertEquals(2, link.abVariants.size)
        assertEquals(1L, link.abVariants[0].id)
        assertEquals("https://example.com/a", link.abVariants[0].targetUrl)
        assertEquals(60, link.abVariants[0].weight)
    }

    @Test
    fun `link defaults the smart-links fields when the server omits them`() = runTest {
        // The legacy SAMPLE_LINK_JSON has none of the smart-links keys.
        server.enqueue(jsonResponse(SAMPLE_LINK_JSON))
        val link = server.client().links.get("q4")
        assertFalse(link.passwordProtected)
        assertNull(link.maxClicks)
        assertEquals(0L, link.clickCount)
        assertFalse(link.trackConversions)
        assertTrue(link.routingRules.isEmpty())
        assertTrue(link.abVariants.isEmpty())
    }

    // ─── create with smart-links input ────────────────────────────────────────

    @Test
    fun `create serializes the smart-links input fields`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_SMART_LINK_JSON))
        server.client().links.create(
            CreateLinkInput(
                targetUrl = "https://example.com/default",
                password = "hunter2",
                maxClicks = 1000,
                trackConversions = true,
                routingRules = listOf(
                    RoutingRule(
                        conditionType = "country",
                        conditionOp = "is",
                        conditionValue = "ZA",
                        targetUrl = "https://example.com/za",
                    ),
                ),
                abVariants = listOf(
                    AbVariantInput(targetUrl = "https://example.com/a", weight = 60),
                    AbVariantInput(targetUrl = "https://example.com/b"),
                ),
            ),
        )
        val body = server.takeRequest().body.readUtf8()
        assertTrue(body.contains("\"password\":\"hunter2\""))
        assertTrue(body.contains("\"max_clicks\":1000"))
        assertTrue(body.contains("\"track_conversions\":true"))
        assertTrue(body.contains("\"condition_type\":\"country\""))
        assertTrue(body.contains("\"condition_op\":\"is\""))
        assertTrue(body.contains("\"condition_value\":\"ZA\""))
        assertTrue(body.contains("\"target_url\":\"https://example.com/za\""))
        assertTrue(body.contains("\"target_url\":\"https://example.com/a\""))
        assertTrue(body.contains("\"weight\":60"))
        // The variant with no weight omits the field.
        assertTrue(body.contains("\"target_url\":\"https://example.com/b\""))
    }

    @Test
    fun `create omits smart-links fields that are not set`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_LINK_JSON))
        server.client().links.create(CreateLinkInput(targetUrl = "https://example.com"))
        val body = server.takeRequest().body.readUtf8()
        assertFalse(body.contains("password"))
        assertFalse(body.contains("max_clicks"))
        assertFalse(body.contains("track_conversions"))
        assertFalse(body.contains("routing_rules"))
        assertFalse(body.contains("ab_variants"))
    }

    // ─── update with smart-links input ────────────────────────────────────────

    @Test
    fun `update serializes routing rules and ab variants as a full replace`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_SMART_LINK_JSON))
        server.client().links.update(
            "vip",
            UpdateLinkInput.builder()
                .trackConversions(true)
                .routingRules(
                    listOf(
                        RoutingRule(
                            conditionType = "device",
                            conditionOp = "is_not",
                            conditionValue = "bot",
                            targetUrl = "https://example.com/human",
                        ),
                    ),
                )
                .abVariants(listOf(AbVariantInput(targetUrl = "https://example.com/v2", weight = 100)))
                .build(),
        )
        val body = server.takeRequest().body.readUtf8()
        assertTrue(body.contains("\"track_conversions\":true"))
        assertTrue(body.contains("\"condition_op\":\"is_not\""))
        assertTrue(body.contains("\"routing_rules\":["))
        assertTrue(body.contains("\"ab_variants\":["))
        assertTrue(body.contains("\"weight\":100"))
    }

    @Test
    fun `update can set a password and a max clicks cap`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_SMART_LINK_JSON))
        server.client().links.update(
            "vip",
            UpdateLinkInput.builder().password("s3cret").maxClicks(500).build(),
        )
        val body = server.takeRequest().body.readUtf8()
        assertTrue(body.contains("\"password\":\"s3cret\""))
        assertTrue(body.contains("\"max_clicks\":500"))
    }

    @Test
    fun `update sends explicit null when clearing password and max clicks`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_SMART_LINK_JSON))
        server.client().links.update(
            "vip",
            UpdateLinkInput.builder().clearPassword().clearMaxClicks().build(),
        )
        val body = server.takeRequest().body.readUtf8()
        assertTrue(body.contains("\"password\":null"))
        assertTrue(body.contains("\"max_clicks\":null"))
    }

    @Test
    fun `update can clear routing rules with an empty list`() = runTest {
        server.enqueue(jsonResponse(SAMPLE_SMART_LINK_JSON))
        server.client().links.update(
            "vip",
            UpdateLinkInput.builder().routingRules(emptyList()).build(),
        )
        val body = server.takeRequest().body.readUtf8()
        assertTrue(body.contains("\"routing_rules\":[]"))
    }

    @Test
    fun `update keeps a real max clicks value over the clear flag`() {
        val input = UpdateLinkInput.builder().maxClicks(750).build()
        assertEquals(750L, input.toJsonMap()["max_clicks"])
    }

    // ─── batch create ─────────────────────────────────────────────────────────

    @Test
    fun `create batch posts the wrapped links body`() = runTest {
        server.enqueue(
            jsonResponse(
                """
                {
                  "created": 2,
                  "total": 3,
                  "results": [
                    {"index": 0, "ok": true, "code": "a1"},
                    {"index": 1, "ok": true, "code": "b2"},
                    {"index": 2, "ok": false, "error": "bad_target_url"}
                  ]
                }
                """,
            ),
        )
        val result = server.client().links.createBatch(
            listOf(
                BatchLinkInput(targetUrl = "https://example.com/1"),
                BatchLinkInput(targetUrl = "https://example.com/2", code = "b2"),
                BatchLinkInput(targetUrl = "nope"),
            ),
        )
        val recorded = server.takeRequest()
        assertEquals("POST", recorded.method)
        assertEquals("/v1/links/batch", recorded.path)
        val body = recorded.body.readUtf8()
        assertTrue(body.contains("\"links\":["))
        assertTrue(body.contains("\"target_url\":\"https://example.com/1\""))
        assertTrue(body.contains("\"code\":\"b2\""))

        assertEquals(2, result.created)
        assertEquals(3, result.total)
        assertEquals(3, result.results.size)
        assertEquals(0, result.results[0].index)
        assertTrue(result.results[0].ok)
        assertEquals("a1", result.results[0].code)
        assertFalse(result.results[2].ok)
        assertEquals("bad_target_url", result.results[2].error)
    }

    @Test
    fun `create batch surfaces a server error`() {
        server.enqueue(jsonResponse("""{"code":"rate_limited","message":"slow down"}""", status = 429))
        val ex = assertThrows(ReroutException::class.java) {
            runTest {
                server.client().links.createBatch(
                    listOf(BatchLinkInput(targetUrl = "https://example.com")),
                )
            }
        }
        assertEquals("rate_limited", ex.code)
        assertTrue(ex.isRateLimited)
    }
}
